mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td

```diff
@@ -260,74 +256,380 @@ def MIOpen_WorkitemIdOp:
+    If the indices into `dest` go out of bounds, the result is undefined.
+  }];
+
+  let assemblyFormat = [{
+    attr-dict $source `->` $dest `[` $coord `]` `:` type($source) `->` type($dest)
+  }];
+  let hasVerifier = 1;
+  let hasCanonicalizeMethod = 1;
+}
+
+// yield
+def MIOpen_YieldOp :
+    MIOpen_Op<"yield", [NoSideEffect, Terminator, ReturnLike]> {
+  let summary = "yield for MIOpen loops";
+
+  let description = [{
+    Op that terminates the MIOpen looping constructs (currently transforming for).
+
+    If the loops produce any values, then this op's arguments should match the types
+    of those values. Otherwise, this op is implicit and can be omitted.
+
+    Note that, in a transforming_for, you don't need to yield back the lower coordinates,
+    as these are managed implicitly.
+  }];
+
+  // Yoinked from affine.yield
+  let arguments = (ins Variadic<AnyType>:$operands);
+
+  let builders = [
+    OpBuilder<(ins), [{ build($_builder, $_state, llvm::None); }]>
+  ];
+
+  let assemblyFormat = "attr-dict ($operands^ `:` type($operands))?";
+}
+
+// Transforming for
+def MIOpen_TransformingForOp :
+    MIOpen_Op<"transforming_for", [RecursiveSideEffects,
+      DeclareOpInterfaceMethods<LoopLikeOpInterface>,
+      AttrSizedOperandSegments,
+      SingleBlockImplicitTerminator<"::mlir::miopen::YieldOp">]>,
+    Arguments<(ins Variadic<Index>:$upperInits,
+      ArgTransformsAttr:$transforms,
+      Variadic<AnyType>:$iterInits,
+      IndexArrayAttr:$bounds,
+      I32ElementsAttr:$lowerStarts,
+      OptionalAttr<UnitAttr>:$useIndexDiffs,
+      OptionalAttr<UnitAttr>:$forceUnroll)>,
+    Results<(outs Variadic<AnyType>:$results)> {
+  let summary = "for loop with coordinate transforms";
+  let description = [{
+    Loops over a rectangular region, using each point as the offset in upper coordinate space(s)
+    before applying coordinate transforms.
+
+    The loop
+    ```mlir
+    %res = miopen.transforming_for {forceUnroll, useIndexDiffs} (%a1, %a2, %a3) = [#transform_map1](%i0, %j0), (%i, %j) = [](%cst0, %cst0)
+      iter_args(%arg0 = %c0 : index) bounds [4, 8] {
+      %v = miopen.buffer_load {...} %place[%a1, %a2, %a3]
+      miopen.buffer_store {...} %v -> %buffer[%i, %j]
+      %cont = addi %arg0, %c1
+      miopen.yield %cont : index
+    }
+    ```
+    will lower to
+    ```mlir
+    %a0, %a1, %a2 = miopen.compute_transforms [#transform_map1](%x, %y)
+    %i, %j = miopen.compute_transforms [](%c0, %c0)
+    %a0_0_0, %a1_0_0 = miopen.index_diff #transform_map1[0, 0] (%x, %y) -> (%a0, %a1, %a2)
+    %i_0_0 = addi %i, %cst0
+    %j_0_0 = addi %j, %cst0
+    %v_0_0 miopen.buffer_load {...} %place[%a0_0_0, %a1_0_0, %a2_0_0]
+    miopen.buffer_store {...} %v -> %buffer[%i_0_0, %j_0_0]
+    %res_0_0 = addi %c0, %c1 : inedx
+    %a0_0_1, %a1_0_1, %a2_0_1 = miopen.index_diff #transform_map1[0, 1] (%x, %y) -> (%a0, %a1, %a2)
+    ...
+    %res_3_7 = addi %res_3_6, %c1 : index
```

Comment:
Several ops in this example don't really exist: `compute_transforms`, `index_diff`. Could this example be revised to depict the proposed semantic?

Also, assuming readers are more familiar with the constructs provided by affine dialect, perhaps it will be better to start from an example without unrolling or index diff maps. And then introduce the effects of unrolling and index diff on top of the naive nested affine loops produced by `transforming_for`.

Reply:
That's what I get for writing the docs first and then changing names partway through. I'll fix it up.